miner/miner.go

```diff
@@ -47,200 +46,150 @@ type Backend interface {
 type Miner struct {
-	mux     *event.TypeMux
-	eth     Backend
-	engine  consensus.Engine
-	exitCh  chan struct{}
-	startCh chan struct{}
-	stopCh  chan struct{}
-	worker  *worker
-
-	wg sync.WaitGroup
-}
+	confMu      sync.RWMutex // The lock used to protect the config
+	config      *Config
+	chainConfig *params.ChainConfig
+	engine      consensus.Engine
+	txpool      *txpool.TxPool
+	chain       *core.BlockChain
 
-func New(eth Backend, config *Config, chainConfig *params.ChainConfig, mux *event.TypeMux, engine consensus.Engine, isLocalBlock func(header *types.Header) bool) *Miner {
-	miner := &Miner{
-		mux:     mux,
-		eth:     eth,
-		engine:  engine,
-		exitCh:  make(chan struct{}),
-		startCh: make(chan struct{}),
-		stopCh:  make(chan struct{}),
-		worker:  newWorker(config, chainConfig, engine, eth, mux, isLocalBlock, true),
-	}
-	miner.wg.Add(1)
-	go miner.update()
-	return miner
-}
+	pendingMu    sync.Mutex // The lock used to protect the pending cache
+	pendingCache *newPayloadResult
+	cacheTime    time.Time
 
-// update keeps track of the downloader events. Please be aware that this is a one shot type of update loop.
-// It's entered once and as soon as `Done` or `Failed` has been broadcasted the events are unregistered and
-// the loop is exited. This to prevent a major security vuln where external parties can DOS you with blocks
-// and halt your mining operation for as long as the DOS continues.
-func (miner *Miner) update() {
-	defer miner.wg.Done()
-
-	events := miner.mux.Subscribe(downloader.StartEvent{}, downloader.DoneEvent{}, downloader.FailedEvent{})
-	defer func() {
-		if !events.Closed() {
-			events.Unsubscribe()
-		}
-	}()
-
-	shouldStart := false
-	canStart := true
-	dlEventCh := events.Chan()
-	for {
-		select {
-		case ev := <-dlEventCh:
-			if ev == nil {
-				// Unsubscription done, stop listening
-				dlEventCh = nil
-				continue
-			}
-			switch ev.Data.(type) {
-			case downloader.StartEvent:
-				wasMining := miner.Mining()
-				miner.worker.stop()
-				canStart = false
-				if wasMining {
-					// Resume mining after sync was finished
-					shouldStart = true
-					log.Info("Mining aborted due to sync")
-				}
-				miner.worker.syncing.Store(true)
-
-			case downloader.FailedEvent:
-				canStart = true
-				if shouldStart {
-					miner.worker.start()
-				}
-				miner.worker.syncing.Store(false)
-
-			case downloader.DoneEvent:
-				canStart = true
-				if shouldStart {
-					miner.worker.start()
-				}
-				miner.worker.syncing.Store(false)
-
-				// Stop reacting to downloader events
-				events.Unsubscribe()
-			}
-		case <-miner.startCh:
-			if canStart {
-				miner.worker.start()
-			}
-			shouldStart = true
-		case <-miner.stopCh:
-			shouldStart = false
-			miner.worker.stop()
-		case <-miner.exitCh:
-			miner.worker.close()
-			return
-		}
-	}
-}
+	// Feeds
+	pendingLogsFeed event.Feed
 
-func (miner *Miner) Start() {
-	miner.startCh <- struct{}{}
+	running atomic.Bool
```

Comment:
I think we can remove this flag, as we have deprecated the background mining notion already